src/backend/postmaster/loginmonitor.c

```diff
@@ -0,0 +1,893 @@
+/*-------------------------------------------------------------------------
+ *
+ * loginmonitor.c
+ *
+ * PostgreSQL Integrated Login Monitor Daemon
+ *
+ * Like autovacuum, the login monitor is structured in two different
+ * kinds of processes: the login monitor launcher and the login monitor
+ * worker. The launcher is an always-running process, started by the
+ * postmaster. It is mainly used to process user's failed login
+ * authentication. It will always running in loop waiting for failed
+ * login signal. The launcher will signal postmaster to fork the worker
+ * process when it receives failed authentication signal from the postgres
+ * process. The worker process is the process which doing the actual
+ * working; as the worker process is only forked when authenticate failed,
+ * one worker process is enough to finish that. It will be forked from
+ * the postmter as needed. Like normal postgres process, login monitor
+ * worker is equipped with locks, transactions, read/write catalog table
+ * and other functionalities.
+ *
+ * The login monitor launcher cannot start the worker process by itself,
+ * as doing so would cause robustness issues (namely, failure to shut
+ * them down on exceptional conditions, and also, since the launcher is
+ * connected to shared memory and is thus subject to corruption there,
+ * it is not as robust as the postmaster). So it leaves that task to the
+ * postmaster.
+ *
+ * There is an login monitor shared memory area, where the launcher
+ * stores information about it's pid and latch. What's more, the worker
+ * stores the current login user's name and it's latch to shared memory.
+ * There is also a flag between launcher and worker to indicate whether
+ * the failed login authentication signal has been processed.
+ *
+ * When there is a failed login authentication, the postgres process will
+ * send the signal to the postmaster and wait the shared memory latch.
+ * The launcher received signal from postgres process, it will set the
+ * flag to true to indicate the signal is processing. And it will resend
+ * a signal to postmaster and wait the latch of lm_latch in shared memory.
+ * After receiving the signal from the launcher, the postmaster will fork
+ * the worker to do the actual working. After the worker finishing working,
+ * it will notify postgres process and the launcher by setting the latch
+ * and the lm_latch in shared memory. Moreover, it will reset the flag to
+ * false to indicate the working is finished.
```

Comment:
`working is finished.`-> `work is finished`

Reply:
working is finished in here is adverbial clause, maybe it's ok.